### PR TITLE
Updated equals and hashcode for MongoCredential

### PR DESCRIPTION
**Pull request by rozza**
userName can now be null so need to null check first.

JAVA-2375

### CODE REVIEW

File: driver-core/src/test/unit/com/mongodb/MongoCredentialSpecification.groovy [hunk 1]
```diff
@@ -27,37 +27,39 @@ import static com.mongodb.AuthenticationMechanism.SCRAM_SHA_1
 class MongoCredentialSpecification extends Specification {
     def 'creating a credential with an unspecified mechanism should populate correct fields'() {
         given:
-        String userName = 'user';
-        String database = 'test';
-        char[] password = 'pwd'.toCharArray();
+        String userName = 'user'
```
Comment (jyemin) [inline driver-core/src/test/unit/com/mongodb/MongoCredentialSpecification.groovy]: The formatting changes are a bit distracting from the main purpose. Can you revert and commit separately?

File: driver-core/src/test/unit/com/mongodb/MongoCredentialSpecification.groovy [hunk 2]
```diff
@@ -192,6 +198,7 @@ class MongoCredentialSpecification extends Specification {
         null == credential.getPassword()
         mechanism == credential.getAuthenticationMechanism()
         MongoCredential.MONGODB_X509_MECHANISM == credential.getMechanism()
+        credential == MongoCredential.createMongoX509Credential()
```
Comment (jyemin) [inline driver-core/src/test/unit/com/mongodb/MongoCredentialSpecification.groovy]: Might as well test hashCode as well.

Comment (jyemin) [inline driver-core/src/test/unit/com/mongodb/MongoCredentialSpecification.groovy]: Given the names of these test methods (... should populate the correct fields), consider creating a separate test method for equals/hashCode ('equivalent credentials should be equal and have same hashCode').

Comment (rozza) [inline driver-core/src/test/unit/com/mongodb/MongoCredentialSpecification.groovy]: Will do

**Comment by rozza** [comment]
Updated the spec

File: driver-core/src/test/unit/com/mongodb/MongoCredentialSpecification.groovy [hunk 3]
```diff
@@ -280,4 +280,28 @@ class MongoCredentialSpecification extends Specification {
 
         !credentialOne.toString().contains(password)
     }
+
+    def 'testEqualsAndHashCode'() {
+        expect:
+        credential() == credential()
```
Comment (jyemin) [inline driver-core/src/test/unit/com/mongodb/MongoCredentialSpecification.groovy]: I haven't seen the empty parens before ( 'credential()').  What are they for?

Comment (rozza) [inline driver-core/src/test/unit/com/mongodb/MongoCredentialSpecification.groovy]: By wrapping the creation in a closure `{}` it means each call to the closure will produce a different instance.  This is so we can test the equality properly and not shortcut out with the same instance.

Comment (jyemin) [inline driver-core/src/test/unit/com/mongodb/MongoCredentialSpecification.groovy]: Oh, didn't notice the closures in the where clause.  Clever.

**Pull request closed without merging**